### DISCUSSION TITLE
using an external dll on Windows

### DISCUSSION THREAD

**Discussion by vipertuy** [tbenthompson/cppimport#93] [Q&A]
Hi, I'm trying to use an external library with cppimport. But I only have a dll file, no lib file.
during the link, cppimport is looking for the .lib file, is it possible to link with a dll file ?

Regards,

**Comment by tbenthompson** [comment]
I suspect something else is going wrong. Can you share the code you're using?

**Reply by vipertuy** [reply]
yes, here it is. I try to use the Fastnoise2 library in this example : 

    // cppimport
    #include <pybind11/pybind11.h>
    #include <Fastnoise/Fastnoise.h>
    
    namespace py = pybind11;
    
    std::vector<float> noise() {
        FastNoise::SmartNode<> fnGenerator = FastNoise::NewFromEncodedNodeTree( "DQAFAAAAAAAAQAgAAAAAAD8AAAAAAA==" );
        std::vector<float> noiseOutput(16 * 16 * 16);
        fnGenerator->GenUniformGrid3D(noiseOutput.data(), 0, 0, 0, 16, 16, 16, 0.2f, 1337);
        return noiseOutput;
    }
    
    PYBIND11_MODULE(somecode, m) {
        m.def("noise", &noise);
    }
    
    <%
    cfg['include_dirs'] = ['../fastnoise/include']
    cfg['libraries'] = ['Fastnoise']
    cfg['library_dirs'] = ['../fastnoise']
    setup_pybind11(cfg)
    %>

and the python code :

  import cppimport.import_hook
  import somecode
  noisedata = noise()

and the error on linking : LINK : fatal error LNK1181: impossible d'ouvrir le fichier en entrée 'Fastnoise.lib' (I'm french !)

**Reply by tbenthompson** [reply]
Hmm, the code looks fine to me. 

I'm guessing the problem is somewhere in your build toolchain. I haven't used Windows myself in over 15 years so I'm not going to be super helpful but this post seems like it might be useful to you: https://stackoverflow.com/a/57109148

I think it'd be a good idea to try to build a C or C++ Python extension without using cppimport. That removes one layer of abstraction and will help debug the issue.

**Reply by vipertuy** [reply]
ok thank you for your answer !
I'll try to work on it.

**Reply by vipertuy** [reply]
Hi, I managed to get the .lib file of FastNoise2, now the linking is allright.
But I have an error on importing somecode in my python code : ImportError: DLL load failed while importing somecode: Le module spécifié est introuvable. Translation = the specified module can't be found
Despite my research, I can't find an issue...

**Reply by tbenthompson** [reply]
I still think you'd have the best luck by starting with a working C or C++ Python binding that does *not* use cppimport. There are plenty of tutorials online that do that. Then, once you've got that working, that will mean that your build toolchain is functioning correctly. 

For this specific issue, my guess is:
- when you're compiling and linking a binary, you need to specify the location of the libraries (including .lib files)
- then later, when you're running that binary (at "runtime"), those libraries also need to be available. Different tools have different flags for indicating where to find binaries at runtime. In Linux, you'd normally use $LIBRARY_PATH. (https://stackoverflow.com/questions/4250624/ld-library-path-vs-library-path) But there will be a different approach in Windows and I'm not familiar with it.
